src/Listener/ServiceListener.php

```diff
@@ -302,4 +291,79 @@ protected function serviceConfigToArray($config)
 
         return $config->toArray();
     }
+
+    /**
+     * Configure the service manager.
+     *
+     * If we've indicated we want to set the application service manager config
+     * metadata, then we will see if any was provided in the merged
+     * configuration, and merge it with any default service configuration we
+     * have in the instance to return a new service manager instance.
+     *
+     * @param array $config
+     * @return ServiceManager
+     */
+    private function configureServiceManager(array $config)
+    {
+        if (! isset($this->serviceManagers[self::IS_APP_MANAGER])) {
+            return $this->defaultServiceManager->withConfig(['services' => [
+                'config' => $config,
+            ]]);
+        }
+
+        $services  = $this->defaultServiceManager;
+        $metadata  = $this->serviceManagers[self::IS_APP_MANAGER];
+
+        $serviceConfig = $this->mergeServiceConfiguration(self::IS_APP_MANAGER, $metadata, $config);
+        $serviceConfig['services']['config'] = $config;
+
+        return $services->withConfig($serviceConfig);
+    }
+
+    /**
+     * Merge all configuration for a given service manager to a single array.
+     *
+     * @param string $key Named service manager
+     * @param array $metadata Service manager metadata
+     * @param array $config Merged configuration
+     * @return array Service manager-specific configuration
+     */
+    private function mergeServiceConfiguration($key, array $metadata, array $config)
+    {
+        if (isset($config[$metadata['config_key']])
```

Comment:
In the future, since the meta data is utilized so heavily and the keys absolutely matter here we should look at turning this into a first class citizen.  Aka this comes internally so it is not an issue per say but we should be designing to a contract in this case: https://github.com/weierophinney/zend-modulemanager/blob/hotfix/mvc-init-integration/src/Listener/ServiceListener.php#L102